tools/bcr_validation.py

```diff
@@ -119,6 +121,56 @@ def fix_line_endings(lines):
     return [line.rstrip() + "\n" for line in lines]
 
 
+def extract_commit(repo_path, path):
+    """
+    Extracts the commit ID from a path matching the pattern /<repo_path>/archive/<commit>.zip or /<repo_path>/archive/<commit>.tar.gz
+
+    Args:
+        repo_path: The repository path.
+        path: The path to extract the commit ID from.
+
+    Returns:
+        The commit ID if found, otherwise None.
+    """
+    pattern = rf"/{re.escape(repo_path)}/archive/([a-f0-9]+)\.(zip|tar\.gz)"
```

Comment:
Assuming malicious spoofing, wouldn't the contributor also be able to game this regex, e.g. by requesting an archive we forgot or inserting double slashes etc.? If we want to make this more reliable, we should probably allowlist the release artifact endpoint and otherwise always run this logic.

Reply:
Or maybe we call `os.path.abspath(parts.path)` first? Which should resolve the path to a normalised form. We do already use this for checking the repo_path.

Reply:
That's another approach, but then we should also run this check for anything after `/archive/`, regardless of hash type, casing or extension.

Reply:
> requesting an archive we forgot or inserting double slashes

Could you explain a bit why these are problems? What does "an archive we forgot" mean?

anyhow, we probably need `^` and `$` surrounding the regex

Reply:
Sorry, this should have read "archive format we forgot". It looks like ZIP and gzipped tars are the only formats GitHub currently supports, but this could change without us noticing and without triggering the check. 

The commit sha can also be specified with uppercase characters.

The check would probably be more reliable if we reuse the source repo check logic and subject everything to the `isSpoofed` check that isn't of the form `https://github.com/{verified source repo}/releases/` after cleaning the path.

Reply:
> subject everything to the isSpoofed check 

This is a great idea, and I did check the API do work for non-commit references. Updated!

Reply:
I'm only allowing extensions ".tar.gz" and ".zip", if github adds support to other extensions, we can extend the list.